db: cache objects with ttl

gioelecerati opened the pull request:
<!-------------------------------------------------------------------------
 | Thanks for send a pull request! 🎉
 | First, please make sure you familiar with the contribution guidelines
 | https://github.com/livepeer/livepeer.studio/blob/master/CONTRIBUTING.md
 -------------------------------------------------------------------------->

**What does this pull request do? Explain your changes. (required)**

Create a cache to store DB objects and save us from spikes of queries

**Specific updates (required)**

<!--- List out all significant updates your code introduces -->

**How did you test each of these updates (required)**

<!-- A detailed description of how you tested your code changes. Include details of your testing environment, and the tests you ran to see how your change affects other areas of the code, etc. -->

**Does this pull request close any open issues?**

<!-- Fixes # -->

**Screenshots (optional)**

<!-- Drag some screenshots here, if applicable -->

**Checklist**

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] I have read the **CONTRIBUTING** document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.

gioelecerati committed:
db: cache objects with ttl

pwilczynskiclearcode commented on packages/api/src/store/cache.ts:
```diff
@@ -0,0 +1,23 @@
+import NodeCache from "node-cache";
+
+export const cache = new NodeCache({ stdTTL: 60, maxKeys: 1000 });
```
Should we document it somewhere that changes to user access may take 1 minute to have effect?

victorges replied:
I think a little read after write inconsistency is somewhat expected in APIs like this (and we also already had _some_ of it due to database read replicas). Since we are trying to only cache objects that are mostly immutable right now (users, object stores, api&access keys, etc) I'd rather keep that behavior implicit for now. 

Wouldn't oppose to document it if anyone has a strong opinion about it though.

victorges committed:
api: Make sure cache objects are not shared

victorges committed:
api/cache: Allow custom ttl for cache

Also remove maxKeys which is bad (it throws an
error, no LRU etc).

victorges committed:
api/playback: Add cache to playback

victorges committed:
api: Make cache gets built-in on db

victorges committed:
api/playback: Cache object store for assets

victorges committed:
api/playback: Fetch streams before assets

We got huge traffic now, lets just switch
their order. Later we can think of optimizing
this better idk

victorges committed:
api: Fix GetOptions default values

useReplica was defaulting to false when the cache: true
option was set.

gioelecerati commented on packages/api/src/middleware/auth.ts:
```diff
@@ -118,10 +119,13 @@ function authenticator(): RequestHandler {
 
-    req.user = user;
+    // make copies in case someone mutates these
+    req.token = { ...tokenObject };
```
This is causing several tests to fail with `access disallowed to api keys`

victorges replied:
Hmmm

victorges replied:
Understood it! problem was when we didnt have an API key here, we were still setting the `token` var to an empty object. Fixed, good catch.

gioelecerati commented on packages/api/src/middleware/auth.ts:
```diff
@@ -108,7 +108,8 @@ function authenticator(): RequestHandler {
     }
 
-    user = await db.user.get(userId);
+    user = await db.user.get(userId, { cache: true });
```
We got a problem here - tests are failing when checking if playback is disallowed to suspended users because the cached user is not suspended

gioelecerati replied:
Also, some other tests are failing due to the User being unverified. We are relying too much to users not being cached

victorges replied:
Yeah we need to update the tests either not to cache at all or to wait until it's expired before checking

victorges replied:
Oh a better option should be a test function to clear the cache

victorges replied:
done

victorges committed:
api: Fix tests

victorges committed:
api: make the cache into a class

victorges committed:
api: Make sure to cache objectstore queries

victorges committed:
api/db: Improve caching logic

- Simplify getOrSet on read
- Always write to cache even if not reading from cache
- Update cache on writes (helps tests more than prod)
- Make copies of objects when reading and writing to cache

victorges committed:
api: Remove unnecessary cache flushes on tests

victorges committed:
api/acl: Make sure all async flows have cache

gioelecerati commented on packages/api/src/controllers/access-control.ts:
```diff
@@ -136,7 +136,7 @@ app.post(
 
     if (user.createdAt < HACKER_DISABLE_CUTOFF_DATE) {
-      let limitReached = await freeTierLimitReached(content, user, req);
+      let limitReached = freeTierLimitReached(content, user, req);
```
Oh! right, wasn't needed!

victorges reviewed (approved):
LGTM 🥸

victorges committed:
api: Remove unused import

victorges committed:
api/cache: Make cache TTL configurable
